gh-122353: Handle `ValueError` during imports (#122389)
python/cpython: Doc/library/importlib.rst

        Loaders that have a file-like storage back-end
        that allows storing arbitrary data
        can implement this abstract method to give direct access
        to the data stored.

>         An :exc:`OSError` is to be raised if the *path* cannot be found, and

Reviewer: I understand that you have split this in a separate paragraph because the text about exceptions is now larger, but "`The *path* is expected`" does not fit here. It should be either split in a separate paragraph (but then we will have too small paragraphs) or merged with the previous paragraph.

Author: How about putting " The
        *path* is expected to be constructed using a module's :attr:`__file__`
        attribute or an item from a package's :attr:`__path__`." after "to the data stored" and have a single paragraph about exceptions?

Reviewer: Yes, this is what I suggested. Sorry for not been clear.

Author: I've also simplified the other occurrences. No need for telling the reason why we would have a ValueError IMO, it's sufficient to know that the exception can be raised.

        :attr:`__path__` attribute.

        If the *path* cannot be handled, this raises an :exc:`OSError`

>         or a :exc:`ValueError` depending on the reason.

Reviewer 2: We deliberately don't document `ValueError` in most cases because it's assumed to be a possibility whenever you pass in an invalid value. If we start documenting it, we'll end up having to put it on every single function.

I'd rather skip all these doc changes and keep with our usual conventions. It may be surprising at first, but the error does only occur when an invalid value is provided, and not due to OS/environment issues (which are the cause of the OSError, and we document functions that may do something that could result in an OSError because most do not).

Reviewer: Since OSError is already explicitly documented here, I do not see harm in mentioning also ValueError which depends on OS and environment. For example, UnicodeEncodeError (which is a subclass of ValueError) can be raised depending on the locale.

This is also important for end user to know which exceptions they should catch. OSError and ValueError are not raised in normal cases, but they can be raised depending on user data, OS and environment. It is easy to miss ValueError.

Such exceptions like TypeError, MemoryError, RecursionError should not be explicitly documented.

Reviewer 2: It depends on the OS, but not on the _state_ of the OS (what I meant by "environment", as opposed to "environment variables" which are only one aspect of the overall state).

ValueError can be raised practically anywhere. If you aren't sure you're passing a valid value, you always need to handle ValueError. It's in the same category as TypeError, MemoryError and RecursionError (and AttributeError when calling a function, as this is essentially the same as TypeError).

Reviewer 3: Another phrasing issue here is the switch from the imperative "is to be raised" to the descriptive "this raises". It's an abstract method definition giving instructions to subclass authors, so the "is to be raised" phrasing should be kept.

Due to that, I do think it's reasonable to mention `ValueError` explicitly here, specifically so subclass method implementors don't feel obliged to catch it and turn it into `OSError`.

>         If the *path* cannot be handled, this raises an :exc:`OSError`
>         or a :exc:`ValueError` depending on the reason.

Reviewer 3: ```suggestion
        If the *path* cannot be handled, either :exc:`OSError` or :exc:`ValueError`
        is to be raised. In most cases, these will be raised by the underlying
        operating system interfaces rather than directly (e.g. :exc:`OSError`
        would be raised when attempting to access a valid but nonexistent
        filesystem path, while attempting to access a path containing a NULL
        byte would raise :exc:`ValueError`).
```

      Reads *path* as a binary file and returns the bytes from it.

>       If the *path* cannot be handled, this raises an :exc:`OSError`
>       or a :exc:`ValueError` depending on the reason.

Reviewer 3: Exceptions are covered in ResourceLoader abstract method documentation above.

```suggestion
```

Author: Actually, I duplicated the entry because the ABC marked as deprecated since 3.7. As such, I'm not sure whether the ABC should update its documentation or not...

        - ``'size'`` (optional): the size in bytes of the source code.

        Any other keys in the dictionary are ignored, to allow for future

>         extensions. If the *path* cannot be handled, this raises an
>         :exc:`OSError` or a :exc:`ValueError` depending on the reason.

Reviewer 3: ```suggestion
        extensions.
        
        As for :meth:`ResourecLoader.get_data`, :exc:`OSError` or
        :exc:`ValueError` is to be raised if the *path* cannot be handled.
```

        .. deprecated:: 3.3
           This method is deprecated in favour of :meth:`path_stats`.  You don't
           have to implement it, but it is still available for compatibility

>            purposes. If the *path* cannot be handled, this raises an
>            :exc:`OSError` or a :exc:`ValueError` depending on the reason.

Reviewer 3: ```suggestion
           purposes.
```

        .. versionchanged:: 3.4
           Raise :exc:`OSError` by default instead of :exc:`NotImplementedError`.

    .. method:: set_data(path, data)

        Optional abstract method which writes the specified bytes to a file
        path. Any intermediate directories which do not exist are to be created
        automatically.

>         When writing to the path fails by raising an :class:`OSError`
>         or a :class:`ValueError`, the exception is not propagated.

Reviewer 3: This change is not correct, only a lack of write permissions should be silently ignored, everything else should be propagated (despite `importlib` itself being overly broad in the errors it suppresses).

```suggestion
        When writing to the path fails because the path is read-only
        (:const:`errno.EACCES`/:exc:`PermissionError`), do not propagate the
        exception. Other exceptions should still be propagated.
```

Author: Since the tests are failing, should we also protect against `FileExistsError` when writing the data? By suppressing OSError altogether, we actually suppressed `FileExistsError` in free-threaded builds and read-only errors (those with errno 30 but that are not explicitly PermissionError).

It seems that the `FileExistsError` only happens in free-threaded builds of Mac OS but I'm not sure whether it's related or not...
